### PR TITLE
Use sdsAllocSize instead of sdsZmallocSize

### PR DESCRIPTION
**Pull request by poiuj**
sdsAllocSize returns the correct size without consulting the
allocator. Which is much faster than consulting the allocator.
The only exception is SDS_TYPE_5, for which it has to
consult the allocator.

This PR also sets alloc field correctly for embedded string objects.
It assumes that no allocator would allocate a buffer larger
than `259 + sizeof(robj)` for embedded string. We use embedded strings
for strings up to 44 bytes. If this assumption is wrong, the whole
function would require a rewrite. In general case sds type adjustment
might be needed. Such logic should go to sds.c.

### CODE REVIEW

**Review by PingXie — APPROVED** [review]
Yeah this change makes perfect sense to me. Thanks @poiuj

File: src/object.c [hunk 1]
```diff
@@ -99,7 +104,11 @@ robj *createEmbeddedStringObject(const char *ptr, size_t len) {
     o->lru = 0;
 
     sh->len = len;
-    sh->alloc = len;
+    size_t usable = bufsize - (sizeof(robj) + sds_hdrlen + 1);
+    sh->alloc = usable;
+    /* Overflow check. Must not happen as we use embedded strings only
+     * for sds strings that fit into SDS_TYPE_8. */
+    assert(usable == sh->alloc);
```
Comment (ranshid) [inline src/object.c]:
```suggestion
    serverAssert(usable == sh->alloc);
```

**Comment by uriyage** [comment]
In `sdsResize`:
```c
 * The sdsAlloc size will be set to the requested size regardless of the actual
 * allocation size, this is done in order to avoid repeated calls to this
 * function when the caller detects that it has excess space. */
sds sdsResize(sds s, size_t size, int would_regrow) {
```

Isn't it an issue?

**Comment by poiuj** [comment]
@uriyage we need to update the comment as this is no longer correct. It was changed in #476. The invariant is that sdsAllocSize always returns the correct size.

Can you share the code where the caller detects the excess space? We need to think of a different solution for that.

File: src/object.c (hunk 1 above)
Comment (poiuj) [reply src/object.c]: Updated.

**Comment by uriyage** [comment]
> @uriyage we need to update the comment as this is no longer correct. It was changed in #476. The invariant is that sdsAllocSize always returns the correct size.
> 
> Can you share the code where the caller detects the excess space? We need to think of a different solution for that.

@poiuj We do it in `clientsCronResizeQueryBuffer`, However since we no longer check with the allocator, it's probably not an issue anymore if we attempt to resize repeatedly. can you please update the `sdsResize` comment.

**Comment by PingXie** [comment]
Good catch on the stale code comment. 

I am not sure I understand the concern about the impact of this PR on the excessive space check logic in `clientsCronResizeQueryBuffer`. With PR #476, the decision between using `alloc` vs consulting with the allocator is an efficiency question, as opposed to correctness, correct? Did I miss anything?

**Comment by poiuj** [comment]
Updated the `sdsResize` comment.

`sdsResize` still calls `je_nallocx` to determine if reallocation makes sense. I'm not sure how costly is this call in context of `clientsCronResizeQueryBuffer`. We can discuss separately if it's something that needs to be improved.

File: src/object.c [hunk 2]
```diff
@@ -99,7 +104,11 @@ robj *createEmbeddedStringObject(const char *ptr, size_t len) {
     o->lru = 0;
 
     sh->len = len;
-    sh->alloc = len;
+    size_t usable = bufsize - (sizeof(robj) + sds_hdrlen + 1);
+    sh->alloc = usable;
+    /* Overflow check. Must not happen as we use embedded strings only
```
Comment (madolson) [inline src/object.c]:
```suggestion
    /* Overflow check. This must not happen as we use embedded strings only
```
Just commenting to retrigger CI.

**Comment by madolson** [comment]
Overall this seems sane to me. I'm running some extra tests since we had some issues last time on the systems that don't use jemalloc.

**Review by madolson — APPROVED** [review]

**Review by lilpoozie2005 — APPROVED** [review]

**Pull request merged**